### ISSUE TITLE
Package request: JReleaser

### ISSUE THREAD

**Issue by msgilligan** [NixOS/nixpkgs#303400]
**Project description**

JReleaser is a Java packaging and release tool.

**Metadata**

* homepage URL: https://jreleaser.org
* source URL: https://github.com/jreleaser/jreleaser
* license: Apache 2.0
* platforms: linux, darwin

**Comment by I-Al-Istannen** [comment]
I have tried this here https://github.com/NixOS/nixpkgs/pull/301914 a while ago simplify the spoon nix flake :^)

Packaging the jar file works and the PR should be fine, but I had some blockers building from source. Getting jreleaser to build *deterministically* is a challenge I wasn't able to overcome so far without sacrificing revision information. JReleaser decided to bake in the date and revision of the last git commit into its binary. Therefore, to properly package jreleaser we need to check out the *git* repository. Sadly, the `fetchgit` or `fetchFromGitHub` helpers are *not deterministic* if you keep `.git`: The object layout is subject to change.

The only real solution I see is cloning without .git and making *my own git commit*. Sadly, users will then see *a revision that does not exist* in their `--version` output, which I consider to be *bad*.

This issue is therefore blocked on upstream I'd say (or nixpkgs maintainers are fine with merging my PR packaging the jar file). If upstream doesn't care about random hashes ending up in their revision that would also be a possible solution.

**Comment by msgilligan** [comment]
⬆️ @aalmiray

**Comment by aalmiray** [comment]
Understood. Does creating the package require strict building from source?

We publish SLSA provenance, SBOMs, and PGP signatures alongside our release artifacts that may be used to verify the authenticity and origin of said artifacts. Also, the generic `.zip` artifact should be reproducible as well.

**Comment by I-Al-Istannen** [comment]
The generic ZIP artifact does not contain .git either. A normal clone with the usual helpers *without* `leaveDotGit` is reproducible too (as that is how git works), but then the revision and build dates are wrong.
The nix package also verifies the hash of the downloaded jar file, so I don't think the additional files are of any real consequence. If anything changes in the jar file the nix package will fail to build.

Packaging from source is not required for Nix (but it is nicer), and my PR currently packages [your release jar file](https://github.com/NixOS/nixpkgs/pull/301914/files#diff-69cc12428101529c45a0f197fbe8007dd0cdaa86e38860c2e8c46ab0f8ea6f24R7). That PR should be merge-able, quite a few java packages only package release jar files. As you don't seem to have a nice other idea for building it from source either, I guess that approach is our best bet at the moment.

**Comment by aalmiray** [comment]
Tis why we provide convenience binaries alongside metadata for verifying their origin and integrity, so that others may consume them directly and skip building from source when such option exists. 

Whether the single JAR (jreleaser-tool-provider) or the generic ZIP is consumed is up to a preference, as the latter contains additional files such as licenses and launcher scripts.

**Comment by I-Al-Istannen** [comment]
Nix packages are normally built from source by [Hydra](https://hydra.nixos.org/build/196107287/download/1/hydra/) and then cached by it. This allows users to retrieve binaries directly (from a trusted cache, such as hydra) and *fall back* to building from source if they need to. Typically this happens when you try to build e.g. old software, build for a different architecture that is not yet built automatically, or just want to be extra safe and not trust any cache. It's sometimes quite useful that many parts can evaluate from source :)

I currently create a wrapper for `jreleaser-cli`, which I think is all the wrapper scripts you need to run the software? I also marked the package as being Apache 2.0 in the meta tags, so I think it should be fine to exclude the LICENSE file itself.

**Comment by I-Al-Istannen** [comment]
This was merged in https://github.com/NixOS/nixpkgs/pull/301914 which made its way to nixpkgs unstable :) It does not build from source (as reproducibly packaging jreleaser without screwing up the version information is something that still eludes me after 5+ hours of work and likely requires patches to plugin dependencies of jreleaser), but it works :)

```
❯ nix run nixpkgs#jreleaser-cli -- --version
------------------------------------------------------------
jreleaser 1.13.1
------------------------------------------------------------
Build time:   2024-06-30 12:32:42Z
Revision:     5542956c0a386c559a22840e3d323d6b59ab9f65
JVM:          21.0.3 (N/A 21.0.3+9-nixos)
------------------------------------------------------------
```

**Comment by tomodachi94** [comment]
Closing; this appears to have been resolved in #301914. Please open a new PR if you encounter any issues with the Nixpkgs package.
